enhancements/sig-policy/98-long-term-compliance-history/README.md

```diff
@@ -0,0 +1,508 @@
+and queryable compliance data.
+
+### Goals
+
+1. Provide an accurate compliance history on the hub or other centralized location.
+1. The compliance history is stored in a queryable format.
+1. The complaince history can contain data for at least 2 years for a hub that manages 3000 managed clusters with 30
+   policies that mostly stay compliant.
+1. The compliance history retains data of deleted managed clusters and deleted policies.
+1. The outcome is that policy compliance events can be queried by a date range and at least the cluster name, parent
+   policy name, policy name, compliance state, and compliance message can be shown. The solution should allow extending
+   with more data.
+1. This feature should be opt-in.
+
+### Non-Goals
+
+1. Open Cluster Management will not manage the Postgresql database. The user must bring their own.
+
+## Proposal
+
+### Design
+
+#### Database Design
+
+The compliance history needs to be stored efficiently in a queryable format. Since the data format is consistent, using
+a relational database is a good fit. A proposed database schema is shown below. Note that this schema allows the same
+database to be used for multiple Open Cluster Management hubs, though if managed cluster names overlap, it'd be less
+confusing if a separate database on the same Postgresql database be used.
+
+##### clusters Table
+
+| Column     | Description                                                                                |
+| ---------- | ------------------------------------------------------------------------------------------ |
+| id         | an internal primary key                                                                    |
+| name       | a string column for the cluster name in the ManagedCluster object; this will have an index |
+| cluster_id | the ACM generated unique cluster ID; this will have a unique constraint                    |
+
+##### parent_policies Table
+
+| Column     | Description                                                         |
+| ---------- | ------------------------------------------------------------------- |
+| id         | an internal primary key                                             |
+| name       | a string column for the name of the parent replicated policy        |
+| categories | an array column of the sorted categories associated with the policy |
+| controls   | an array column of the sorted controls associated with the policy   |
+| standards  | an array column of the sorted standards associated with the policy  |
+
+There would be a combined unique constraint on the `name`, `categories`, `controls`, and `standards` columns. Note that
+the array column is a Postgresql feature and not supported in databases such as SQLite. If we do need to support other
+database types in the future, these can be string columns and perhaps have filtering on the API disabled for these
+columns.
```

Comment:
Do other SQL flavors allow filters like "if string x is found inside this column?" Like a substring match...

The way the policy already stores these is just as a comma-delimited list, since each annotation is a string, not a list. So if we could just put each annotation directly into these columns, that could be slightly neater on that side... and then regardless of the database flavor, it might still be possible to get the filtering.

Reply:
@JustinKuli, I think a JSON column type may be better for portability to another database type. Postgres and MySQL both have it. SQLite has JSON functions that can parse text columns.

Reply:
After doing some more research, I don't think a JSON type is good for unique constraints. I like your idea of allowing substring checking for non-Postgresql databases.